File: day-09/part-1/youyoun.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
	"time"
)

type Program struct {
	intCode    map[int]int
	pointer    int
	pointerRef int
	inputValue int
	outputs    []int
	end        int
}

var OpNArgsMap = map[int]int{
	1:  3,
	2:  3,
	3:  1,
	4:  1,
	5:  2,
	6:  2,
	7:  3,
	8:  3,
	9:  1,
	99: 0,
}

func parseInstruction(instr int) (int, []int) {
	op := instr % 100
	intModes := instr / 100
	modes := []int{0, 0, 0}
	for i := 0; i < 3; i++ {
		modes[i] = intModes % 10
		intModes /= 10
	}
	return op, modes
}

func (p Program) param(pos int, mode int) int {
	switch mode {
	case 0:
		return p.intCode[p.pointer+pos]
	case 1:
		return p.pointer + pos
	case 2:
		return p.pointerRef + p.intCode[p.pointer+pos]
	default:
		panic("Invalid param mode")
	}
}

func (p *Program) runInstructions() {
	instr := p.intCode[p.pointer]
	opcode, modes := parseInstruction(instr)
	switch opcode {
	case 1:
		p.intCode[p.param(3, modes[2])] = p.intCode[p.param(2, modes[1])] + p.intCode[p.param(1, modes[0])]
	case 2:
		p.intCode[p.param(3, modes[2])] = p.intCode[p.param(2, modes[1])] * p.intCode[p.param(1, modes[0])]
	case 3:
		p.intCode[p.param(1, modes[0])] = p.inputValue
	case 4:
		p.outputs = append(p.outputs, p.intCode[p.param(1, modes[0])])
	case 5:
		if p.intCode[p.param(1, modes[0])] != 0 {
			p.pointer = p.intCode[p.param(2, modes[1])]
			return
		}
	case 6:
		if p.intCode[p.param(1, modes[0])] == 0 {
			p.pointer = p.intCode[p.param(2, modes[1])]
			return
		}
	case 7:
		if p.intCode[p.param(1, modes[0])] < p.intCode[p.param(2, modes[1])] {
			p.intCode[p.param(3, modes[2])] = 1
		} else {
			p.intCode[p.param(3, modes[2])] = 0
		}
	case 8:
		if p.intCode[p.param(1, modes[0])] == p.intCode[p.param(2, modes[1])] {
			p.intCode[p.param(3, modes[2])] = 1
		} else {
			p.intCode[p.param(3, modes[2])] = 0
		}
	case 9:
		p.pointerRef += p.intCode[p.param(1, modes[0])]
	case 99:
		p.end = 1
		return
	default:
		panic("Invalid OP code")
	}
	p.pointer += OpNArgsMap[opcode] + 1
}

func run(s string) interface{} {
	intCode := make(map[int]int)
	for ind, val := range strings.Split(s, ",") {
		intCode[ind], _ = strconv.Atoi(val)
	}
	prog := Program{
		intCode:    intCode,
		inputValue: 1,
	}
	for {
		prog.runInstructions()
		if prog.end == 1 {
			break
		}
	}
	return prog.outputs[0]
}

func main() {
	// Uncomment this line to disable garbage collection
	// debug.SetGCPercent(-1)

	// Read input from stdin
	input, err := ioutil.ReadAll(os.Stdin)
	if err != nil {
		panic(err)
	}

	// Start resolution
	start := time.Now()
	result := run(string(input))

	// Print result
	fmt.Printf("_duration:%f\n", time.Now().Sub(start).Seconds()*1000)
	fmt.Println(result)
}

File: day-09/part-1/youyoun.py
from tool.runners.python import SubmissionPy
from typing import List

MAX_PARAM_NUM = 3


class InfList:
    def __init__(self, list_):
        self.list_ = list_

    def expand(self, value):
        self.list_.extend([0 for _ in range(value - len(self) + 1)])

    def __getitem__(self, item):
        if len(self) <= item:
            self.expand(item)
        return self.list_[item]

    def __setitem__(self, key, value):
        if len(self) <= key:
            self.expand(key)
        self.list_[key] = value

    def __len__(self):
        return len(self.list_)

    def __repr__(self):
        return repr(self.list_)


class Program:
    def __init__(self, intcode: List):
        self.intcode = InfList(intcode)
        self.pos = 0
        self.opcode = 0
        self.input_value = 1
        self.outputs = []
        self.end = 0
        self.pointer_ref = 0

    def set_input(self, inp):
        self.input_value = inp

    def move_pointer(self):
        if self.opcode in [1, 2, 7, 8]:
            self.pos += 4
        elif self.opcode in [5, 6]:
            self.pos += 3
        elif self.opcode in [3, 4, 9]:
            self.pos += 2
        else:
            pass

    def run_instruction(self):
        instruction_code = self.intcode[self.pos]
        self.opcode, param_modes = parse_code(instruction_code)
        # 1002,4,3,4,33
        if self.opcode == 1:
            self.intcode[self.param(self.pos + 3, param_modes[2])] = self.intcode[self.param(self.pos + 1, param_modes[0])] \
                                                        + \
                                                        self.intcode[self.param(self.pos + 2, param_modes[1])]
        elif self.opcode == 2:
            self.intcode[self.param(self.pos + 3, param_modes[2])] = self.intcode[self.param(self.pos + 1, param_modes[0])] \
                                                        * \
                                                        self.intcode[self.param(self.pos + 2, param_modes[1])]
        elif self.opcode == 3:
            self.intcode[self.param(self.pos + 1, param_modes[0])] = self.input_value
        elif self.opcode == 4:
            self.outputs.append(self.intcode[self.param(self.pos + 1, param_modes[0])])
        elif self.opcode == 5:
            if self.intcode[self.param(self.pos + 1, param_modes[0])] != 0:
                self.pos = self.intcode[self.param(self.pos + 2, param_modes[1])]
                return  # dont run self.move_pointer
        elif self.opcode == 6:
            if self.intcode[self.param(self.pos + 1, param_modes[0])] == 0:
                self.pos = self.intcode[self.param(self.pos + 2, param_modes[1])]
                return
        elif self.opcode == 7:
            self.intcode[self.param(self.pos + 3, param_modes[2])] = int(self.intcode[self.param(self.pos + 1, param_modes[0])]
                                                            < self.intcode[self.param(self.pos + 2, param_modes[1])])
        elif self.opcode == 8:
            self.intcode[self.param(self.pos + 3, param_modes[2])] = int(self.intcode[self.param(self.pos + 1, param_modes[0])]
                                                            == self.intcode[self.param(self.pos + 2, param_modes[1])])
        elif self.opcode == 9:
            self.pointer_ref += self.intcode[self.param(self.pos + 1, param_modes[0])]
        elif self.opcode == 99:
            self.end = 1
        else:
            raise ValueError("Invalid op code")
        self.move_pointer()
        return

    def param(self, pos, mode):
        if mode == 0:
            return self.intcode[pos]
        elif mode == 1:
            return pos
        elif mode == 2:
            return self.pointer_ref + self.intcode[pos]
        else:
            raise ValueError("Invalid parameter mode")


def parse_code(opcode):
    operation = opcode % 100
    all_modes = opcode // 100
    param_modes = [0 for _ in range(MAX_PARAM_NUM)]
    for i in range(MAX_PARAM_NUM):
        param_modes[i] = all_modes % 10
        all_modes = all_modes // 10
    return operation, param_modes


class YouyounSubmission(SubmissionPy):

    def run(self, s):
        # :param s: input in string format
        # :return: solution flag
        # Your code goes here
        instructions = [int(x) for x in s.split(",")]
        prog = Program(instructions)
        while prog.end == 0:
            prog.run_instruction()
        return prog.outputs[-1]
